### PR TITLE
Add `ProgrammableAncillaRotation` with costs

### PR DESCRIPTION
**Pull request by anurudhp**
partially addresses #1135

### CODE REVIEW

**Comment by anurudhp** [comment]
cc @tanujkhattar @ncrubin

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+from qualtran import Bloq, bloq_example, BloqBuilder, QBit, Register, Side, Signature, SoquetT
+from qualtran.bloqs.basic_gates import CNOT, Hadamard, Rz, XGate
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.symbolics import ceil, is_symbolic, log2, SymbolicFloat, SymbolicInt
+
+
+@frozen
+class RzResourceState(Bloq):
+    r"""Resource qubit with state $Rz(\phi) |+\rangle$"""
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/programmable_ancilla_rotation.py:29]:
```suggestion
    r"""Resource qubit prepared in state $Rz(\phi) |+\rangle$ = $|0\rangle + e^{i\phi}|1\rangle$"""
```

Although the reference says the prepared state is $|0\rangle + e^{i\phi}|1\rangle$, if you take the definition of $Rz(t) = e^{-i Z t /2}$ as commonly defined, the prepared state would be $e^{-i\phi/2}|0\rangle + e^{i\phi/2}|1\rangle$.  It's worth verifying that the procedure works as expected with the right global phase.

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            resources |= {
+                (
+                    RzResourceState(
+                        2**self.n_rounds * self.angle, eps=self.eps / 2**self.n_rounds
+                    ),
+                    1,
+                )
+            }
+
+        # TODO: account for `n_rounds` measurements
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/rotations/programmable_ancilla_rotation.py]: ah right, added `CirqGateAsBloq(Measure())` for now

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        if self.apply_final_correction:
+            resources |= {
+                (
+                    RzResourceState(
+                        2**self.n_rounds * self.angle, eps=self.eps / 2**self.n_rounds
+                    ),
+                    1,
+                )
+            }
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/rotations/programmable_ancilla_rotation.py]: Oops, you're right, it should be an Rz gate, not resource state. Fixed now

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        if is_symbolic(self.n_rounds):
+            phi = ssa.new_symbol(r"\phi")
+            eps = ssa.new_symbol(r"\epsilon")
+            resources = {(RzResourceState(phi, eps), self.n_rounds)}
+        else:
+            resources = {
+                (RzResourceState(2**i * self.angle, eps=self.eps / 2 ** (i + 1)), 1)
+                for i in range(int(self.n_rounds))
```
[the remark this answers is not in the document]
Comment (anurudhp) [reply qualtran/bloqs/rotations/programmable_ancilla_rotation.py]: Fixed to use eps/n

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py (hunk 1 above)
Comment (anurudhp) [reply qualtran/bloqs/rotations/programmable_ancilla_rotation.py]: fixed: used ZPow instead of Rz

**Comment by anurudhp** [comment]
@tanujkhattar can we merge this?

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 5]
```diff
@@ -72,90 +78,96 @@ class RzViaProgrammableAncillaRotation(Bloq):
...
         if self.apply_final_correction:
-            resources |= {
-                (
-                    RzResourceState(
-                        2**self.n_rounds * self.angle, eps=self.eps / 2**self.n_rounds
-                    ),
-                    1,
-                )
-            }
+            resources[Rz(2**self.n_rounds * self.exponent, eps=self.eps / n_rz)] += 1
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/programmable_ancilla_rotation.py:133]: Should this final correction be an `Rz` or a `ZPow` ?

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    r"""Single qubit ZPow rotation using resource states.
+
+    This bloq applies a single qubit Z**t rotation only using ZPow resource states and
+    clifford gates. It is designed to exit early as soon as a measurement succeeds.
+
+    Notes:
+        - This bloq uses measurements.
+        - To use this Bloq in costing, use the precise number of rounds that are actually
+          expected during execution. As Qualtran does not support analyzing measurement-based
+          post-selection circuits, the complexity of this Bloq is the worst-case for the
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/programmable_ancilla_rotation.py:74]: Please add a reference to https://github.com/quantumlib/Qualtran/issues/445 where we say Qualtran doesn't yet support measurements and classical data.

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    zpow_programmed_ancilla_symb = ZPowProgrammedAncilla(t)
+    return zpow_programmed_ancilla_symb
+
+
+@frozen
+class ZPowUsingProgrammedAncilla(Bloq):
+    r"""Single qubit ZPow rotation using resource states.
+
+    This bloq applies a single qubit Z**t rotation only using ZPow resource states and
+    clifford gates. It is designed to exit early as soon as a measurement succeeds.
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/programmable_ancilla_rotation.py:68]: Please add a brief explanation of the strategy used to apply `Z**t` and the corresponding costs.

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    @classmethod
+    def from_failure_probability(
+        cls,
+        exponent: SymbolicFloat,
+        *,
+        max_fail_probability: SymbolicFloat,
+        eps: SymbolicFloat = 1e-11,
+    ) -> 'ZPowUsingProgrammedAncilla':
+        """Applies the rotation `Rz(angle)` except with some specified failure probability.
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/programmable_ancilla_rotation.py:109]: `Rz(angle)` is stale; please update to `ZPow`

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
```
Comment (tanujkhattar) [inline qualtran/bloqs/rotations/programmable_ancilla_rotation.py:110]: Add an explanation of the formula used to compute the number of rounds using failure probability.

**Review by tanujkhattar — APPROVED** [review]
LGTM % docstrings.

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py (hunk 8 above)
Comment (anurudhp) [reply qualtran/bloqs/rotations/programmable_ancilla_rotation.py]: fixed

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py (hunk 6 above)
Comment (anurudhp) [reply qualtran/bloqs/rotations/programmable_ancilla_rotation.py:83]: done

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py (hunk 5 above)
Comment (anurudhp) [reply qualtran/bloqs/rotations/programmable_ancilla_rotation.py]: `ZPow`, fixed

File: qualtran/bloqs/rotations/programmable_ancilla_rotation.py (hunk 9 above)
Comment (anurudhp) [reply qualtran/bloqs/rotations/programmable_ancilla_rotation.py:123]: added

**Pull request merged**